cborg/src/Codec/CBOR/Decoding.hs

```diff
@@ -131,13 +136,15 @@ data DecodeAction s a
     | ConsumeTag64     (Word64# -> ST s (DecodeAction s a))
 #endif
 
-    | ConsumeInteger (Integer    -> ST s (DecodeAction s a))
-    | ConsumeFloat   (Float#     -> ST s (DecodeAction s a))
-    | ConsumeDouble  (Double#    -> ST s (DecodeAction s a))
-    | ConsumeBytes   (ByteString -> ST s (DecodeAction s a))
-    | ConsumeString  (Text       -> ST s (DecodeAction s a))
-    | ConsumeBool    (Bool       -> ST s (DecodeAction s a))
-    | ConsumeSimple  (Word#      -> ST s (DecodeAction s a))
+    | ConsumeInteger       (Integer   -> ST s (DecodeAction s a))
+    | ConsumeFloat         (Float#    -> ST s (DecodeAction s a))
+    | ConsumeDouble        (Double#   -> ST s (DecodeAction s a))
+    | ConsumeBytes         (ByteString-> ST s (DecodeAction s a))
+    | ConsumeByteArray     (BA.ByteArray -> ST s (DecodeAction s a))
+    | ConsumeString        (Text      -> ST s (DecodeAction s a))
+    | ConsumeUtf8ByteArray (BA.ByteArray -> ST s (DecodeAction s a))
```

Comment:
For all these interpreter instructions we should be clear, and document below:

 * are they copying or non-copying
 * for UTF8 are they validating or non-validating.

If we're promising copying (ie no retention of input) then we really ought to have tests to check this property. I think we currently violate it for `ConsumeBytes`. I have a patch sitting somewhere to insert a `BS.copy`.